### PR TITLE
Fix #504

### LINKED ISSUE #504: Conditional Parameters in Trace Serialization
Given Hierarchical parameters, a random search like procedure would not always want to go through the burden of specifying values for all hyperparameters (in fact, randomsearchcv does, but one of my spinoff classes, called defaultsearch, does not. And I think for a good reason, although we can discuss the validity of that)

However, when running the following piece of code:
```
    def test_run_and_upload_defaultsearch(self):
        # more complicated, as it can have hierarchical parameters
        defaultsearch = DefaultSearchCV(
            RandomForestClassifier(n_estimators=5),
            [{'max_features': 4}, {'min_samples_leaf': 1}],
            cv=StratifiedKFold(n_splits=2, shuffle=True))
        # The random states for the RandomizedSearchCV is set after the
        # random state of the RandomForestClassifier is set, therefore,
        # it has a different value than the other examples before
        self._run_and_upload(defaultsearch, '12172')
```
I get the following error:
```
Error
Traceback (most recent call last):
  File "/home/jan/anaconda3/envs/openml-defaults/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/home/jan/anaconda3/envs/openml-defaults/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/jan/projects/openml-python/tests/test_runs/test_run_functions.py", line 421, in test_run_and_upload_defaultsearch
    self._run_and_upload(defaultsearch, '12172')
  File "/home/jan/projects/openml-python/tests/test_runs/test_run_functions.py", line 340, in _run_and_upload
    random_state_value=rsv)
  File "/home/jan/projects/openml-python/tests/test_runs/test_run_functions.py", line 154, in _perform_run
    avoid_duplicate_runs=openml.config.avoid_duplicate_runs)
  File "/home/jan/projects/openml-python/openml/runs/functions.py", line 120, in run_flow_on_task
    res = _run_task_get_arffcontent(flow.model, task, add_local_measures=add_local_measures)
  File "/home/jan/projects/openml-python/openml/runs/functions.py", line 454, in _run_task_get_arffcontent
    add_local_measures=add_local_measures)
  File "/home/jan/projects/openml-python/openml/runs/functions.py", line 576, in _run_model_on_fold
    arff_tracecontent.extend(_extract_arfftrace(model, rep_no, fold_no))
  File "/home/jan/projects/openml-python/openml/runs/functions.py", line 644, in _extract_arfftrace
    serialized_value = json.dumps(model.cv_results_[key][itt_no])
  File "/home/jan/anaconda3/envs/openml-defaults/lib/python3.6/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/home/jan/anaconda3/envs/openml-defaults/lib/python3.6/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/home/jan/anaconda3/envs/openml-defaults/lib/python3.6/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/home/jan/anaconda3/envs/openml-defaults/lib/python3.6/json/encoder.py", line 180, in default
    o.__class__.__name__)
TypeError: Object of type 'MaskedConstant' is not JSON serializable
```
Because the MaskedArray cv_result_ is now a 2x2 array, with the empty indices as MaskedConstant. I think we should add an if else statement checking for this in the trace serialization. WDYT?

### PR DESCRIPTION
**Pull request by janvanrijn**
fix proposal for #504

### CODE REVIEW

**Comment by janvanrijn** [comment]
@mfeurer can you please review this one?

**Comment by janvanrijn** [comment]
@mfeurer is there any reason to not merge this one yet?

**Comment by janvanrijn** [comment]
@mfeurer ping

File: openml/flows/sklearn_converter.py [hunk 1]
```diff
@@ -625,28 +625,33 @@ def _serialize_cross_validator(o):
 
 
 def _check_n_jobs(model):
```
Comment (mfeurer) [inline openml/flows/sklearn_converter.py]: Given the additional duties of this function, could you please change the function name?

File: openml/flows/sklearn_converter.py [hunk 2]
```diff
@@ -625,28 +625,33 @@ def _serialize_cross_validator(o):
...
-    '''
-    def check(param_dict, disallow_parameter=False):
-        for param, value in param_dict.items():
-            # n_jobs is scikitlearn parameter for paralizing jobs
-            if param.split('__')[-1] == 'n_jobs':
-                # 0 = illegal value (?), 1 = use one core,  n = use n cores
-                # -1 = use all available cores -> this makes it hard to
-                # measure runtime in a fair way
-                if value != 1 or disallow_parameter:
+    will run on a single core (in that case, openml-python can measure runtimes)
```
Comment (mfeurer) [inline openml/flows/sklearn_converter.py]: as well as the function documentation?

File: openml/runs/trace.py [hunk 3]
```diff
@@ -117,10 +117,10 @@ def get_parameters(self):
         result = {}
         # parameters have prefix 'parameter_'
         prefix = 'parameter_'
-
         for param in self.setup_string:
             key = param[len(prefix):]
-            result[key] = json.loads(self.setup_string[param])
+            value = self.setup_string[param]
+            result[key] = json.loads(value)
```
Comment (mfeurer) [inline openml/runs/trace.py]: Is there a reason you touched this file? If not, could you please revert these changes to keep this commit focused?

File: tests/test_runs/test_run_functions.py [hunk 4]
```diff
@@ -98,12 +98,17 @@ def _check_serialized_optimized_run(self, run_id):
...
             for col_idx in compare_slice:
-                self.assertEquals(predictions['data'][idx][col_idx], predictions_prime['data'][idx][col_idx])
+                self.assertEqual(predictions['data'][idx][col_idx], predictions_prime['data'][idx][col_idx])
 
         return True
 
     def _perform_run(self, task_id, num_instances, clf,
                      random_state_value=None, check_setup=True):
+        # last case (DefaultSearchCV) relies on where the test is invoked from
+        # TODO: move it to a 'permanent' place?
```
Comment (mfeurer) [inline tests/test_runs/test_run_functions.py]: I'm having trouble understanding this comment. Could you please clarify?

File: tests/test_runs/test_run_functions.py [hunk 5]
```diff
@@ -374,6 +388,23 @@ def test_run_and_upload_randomsearch(self):
         # it has a different value than the other examples before
         self._run_and_upload(randomsearch, '12172')
 
+    def test_run_and_upload_maskedarrays(self):
+        # This testcase is important for 2 reasons:
+        # 1) it verifies the correct handling of masked arrays (not all parameters are active)
+        # 2) it verifies the correct handling of a 2-layered grid search
+        gridsearch = GridSearchCV(
```
Comment (mfeurer) [inline tests/test_runs/test_run_functions.py]: Could you please explain what makes this example conditional? Right now it looks like a shorter version of the example above and I don't see where the conditionality comes in.

File: openml/flows/sklearn_converter.py (hunk 1 above)
Comment (janvanrijn) [reply openml/flows/sklearn_converter.py]: There are no additional duties, just a refactoring (to make subfunction more readible) and the correct assumptions of the param_grid (which can be of datatype dict and list of dicts)

File: openml/flows/sklearn_converter.py (hunk 2 above)
Comment (janvanrijn) [reply openml/flows/sklearn_converter.py]: Same. Function description is still up to date.

File: openml/runs/trace.py (hunk 3 above)
Comment (janvanrijn) [reply openml/runs/trace.py]: Yes, I changed this to make it easier to inspect `value`. Due to reasons I forgot the json loader crashed, this issue is resolved, but it seems good to keep it like this.

File: tests/test_runs/test_run_functions.py (hunk 5 above)
Comment (janvanrijn) [reply tests/test_runs/test_run_functions.py]: I added "Note that this is a list of dictionaries, all containing 1 hyperparameter. " I assume you agree

**Comment by janvanrijn** [comment]
@mfeurer I think we are good to merge
